### PR TITLE
Support for HTTP Basic authentication

### PR DESCRIPTION
**Pull request by dynamind**
HTTP Basic authentication can be disabled and is enabled by default. This PR includes updated documentation.

Functional tests need to be run twice (once for each client authentication method), which is controlled by setting the system property `http.basic` to `true`, e.g. `grails test-app functional: -Dhttp.basic=true`.

I've aimed to keep it as clean and simple as possible. Extra effort went into ensuring that various exceptions were handled in a manner identical to the previous mechanism.

### CODE REVIEW

File: SpringSecurityOauth2ProviderGrailsPlugin.groovy [hunk 1]
```diff
@@ -470,6 +479,41 @@ class SpringSecurityOauth2ProviderGrailsPlugin {
...
+    private configureBasicAuthenticationFilter = { conf ->
+        // Should the basic authentication filter be registered in the filter chain
+        boolean registerBasicAuthFilter = conf.oauthProvider.registerBasicAuthenticationFilter as boolean
+
+        if(registerBasicAuthFilter) {
+            // Create a separate authentication manager with a single provider for the basic authentication filter
+            basicClientAuthenticationManager(ProviderManager) {
+                providers = [ ref('clientCredentialsAuthenticationProvider') ]
+                authenticationEventPublisher = ref('authenticationEventPublisher')
+                eraseCredentialsAfterAuthentication = conf.providerManager.eraseCredentialsAfterAuthentication // true
```
Comment (bobby-vandiver) [inline SpringSecurityOauth2ProviderGrailsPlugin.groovy]: Is there ever a case where this shouldn't be `true`?

**Comment by bobby-vandiver** [comment]
Re: the configuration options provided by the Core plugin that we're reusing. We should provide options specific to the OAuth2 plugin so the two configurations can vary. 

Other than that, great work. :+1:

File: SpringSecurityOauth2ProviderGrailsPlugin.groovy (hunk 1 above)
Comment (dynamind) [inline SpringSecurityOauth2ProviderGrailsPlugin.groovy]: I don't believe so. I'll change it to `eraseCredentialsAfterAuthentication = true`

**Comment by dynamind** [comment]
In lieu of the recent issue reports all having something to do with the filter chain configuration, I think it's good to note that the Core plugin's `basicAuthenticationFilter` should not be mixed with our new `oauth2BasicAuthenticationFilter`. In summary:

- Remove `basicAuthenticationFilter` from `/oauth/token`
- Remove both `basicAuthenticationFilter` and `oauth2BasicAuthenticationFilter` from the resource mapping
- Remove `oauth2BasicAuthenticationFilter` from `/**`

I've tested this in one of my own projects and it works as expected.

**Pull request merged**